source/IAccessibleHandler/__init__.py

```diff
@@ -239,6 +244,99 @@
+	)
+
+
+def getStatesSetFromIAccessible2States(IAccessible2States: int) -> Set[State]:
+	return set(
+		IAccessible2StatesToNVDAStates[IA2State]
+		for IA2State in IAccessible2StatesToNVDAStates.keys()
+		if IA2State & IAccessible2States
+	)
+
+
+def getStatesSetFromIAccessibleAttrs(attrs: "textInfos.ControlField") -> Set[State]:
+	# States are serialized (in XML) with an attribute per state.
+	# The value for the state is used in the attribute name.
+	# The attribute value is always 1.
+	# EG IAccessible::state_40="1"
+	IAccessibleStateAttrName = 'IAccessible::state_{}'
+	return set(
+		IAccessibleStatesToNVDAStates[IAState]
+		for IAState in IAccessibleStatesToNVDAStates.keys()
+		if int(attrs.get(IAccessibleStateAttrName.format(IAState), 0))
+	)
+
+
+def getStatesSetFromIAccessible2Attrs(attrs: "textInfos.ControlField") -> Set[State]:
+	# States are serialized (in XML) with an attribute per state.
+	# The value for the state is used in the attribute name.
+	# The attribute value is always 1.
+	# EG IAccessible2::state_40="1"
+	IAccessible2StateAttrName = 'IAccessible2::state_{}'
+	return set(
+		IAccessible2StatesToNVDAStates[IA2State]
+		for IA2State in IAccessible2StatesToNVDAStates.keys()
+		if int(attrs.get(IAccessible2StateAttrName.format(IA2State), 0))
+	)
+
+
+def calculateNvdaRole(IARole: int, IA2States: int) -> Role:
+	NVDARole = IAccessibleRolesToNVDARoles.get(IARole, Role.UNKNOWN)
+	# Don't report indeterminate progress bars as "half-checked"
+	# L{State.HALFCHECKED} maps from oleacc.STATE_SYSTEM_MIXED
+	# which has the same value as oleacc.STATE_SYSTEM_INDETERMINATE
+	# L{State.INDETERMINATE} is not mapped directly from any IA or IA2 state
+	# A progress bar that can not convey progress, only activity, is a busy indicator.
+	if NVDARole == Role.PROGRESSBAR and oleacc.STATE_SYSTEM_INDETERMINATE & IA2States:
+		return Role.BUSY_INDICATOR
+	return NVDARole
+
+
+def calculateNvdaStates(IARole: int, IA2States: int) -> Set[State]:
+	roles = IAccessibleRolesToNVDARoles.get(IARole, Role.UNKNOWN)
+	states = _getStatesSetFromIAccessibleStates(IA2States)
+	# Don't report indeterminate progress bars as "half-checked"
+	# L{State.HALFCHECKED} maps from oleacc.STATE_SYSTEM_MIXED
+	# which has the same value as oleacc.STATE_SYSTEM_INDETERMINATE
+	# L{State.INDETERMINATE} is not mapped directly from any IA or IA2 state
+	# A progress bar that can not convey progress, only activity, is a busy indicator.
+	if roles == Role.PROGRESSBAR and states.intersection({State.HALFCHECKED, State.INDETERMINATE}):
+		states.discard(State.HALFCHECKED)
+		states.add(State.INDETERMINATE)
+		return states
+	return states
+
+
+def NVDARoleFromAttr(accRole: Union[int, str, None]) -> Role:
+	if accRole.isdigit():
+		accRole = int(accRole)
+	else:
+		accRole = accRole.lower()
+	return IAccessibleRolesToNVDARoles.get(accRole, controlTypes.Role.UNKNOWN)
+
+
+def transformRoleStates(role: Role, states: Set[State]) -> Tuple[Role, Set[State]]:

```

Comment:
This function could use some documentation like calculateNvdaStates

Reply:
While writing docs here, I don't think this belongs in `IAccessibleHandler`, it is not specific to IAccessible because it is "post-conversion" and dealing only with NVDA Role and NVDA States.

I'm going to move it.